Add controllers

Rat-In-Hat opened the pull request.

Rat-In-Hat committed:
feat

реализовать класс User, UserMapper и UserDto
реализовать классы хранения, сервиса и контроллера с базовыми операциями над пользователем

Rat-In-Hat committed:
style

code reformat

Rat-In-Hat committed:
feat

реализовать классы хранилища предметов, сервиса и класс контроллера. Пройдены основные тесты постман коллекции, касаемые item

Rat-In-Hat committed:
style

reformat code

Rat-In-Hat committed:
feat

добавлено логирование в контроллеры

Rat-In-Hat committed:
feat

добавлены заготовки для бронирования и запросов для реализации в след спринте

Rat-In-Hat committed:
style

изменить имя перемменой чтобы пройти тесты на гите

FanaticFiz commented on src/main/java/ru/practicum/shareit/booking/dto/BookingDto.java:
```diff
@@ -1,7 +1,21 @@
+@Data
+@Builder
 public class BookingDto {
+    private final Long id; // ID of the booking
```
Это все не используемые заготовки, но тем не менее:
- id у dto не будет
- комментарии вот такого плана не нужны, названия переменных хватает

Rat-In-Hat replied:
Убрал поле id пока только у request и booking, когда убираю у user и item сыпятся тесты постман, так как он ожидает поле id ,  без него увы тесты не идут. Либо оставлять у DTO user , item поле ID, либо возвращать не DTO, а User, Item классы

ненужные комментарии убрал

FanaticFiz commented on src/main/java/ru/practicum/shareit/item/controller/ItemController.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    @PatchMapping("/{itemId}")
+    public ResponseEntity<ItemDto> updateItem(@RequestHeader(requestHeader) Long userId,
```
- Параметры хочется видеть рядом, вот так "причесать" например

```
    public ResponseEntity<ItemDto> updateItem(@RequestHeader(requestHeader) Long userId,
                                              @PathVariable Long itemId,
                                              @RequestBody Item item) {
```

- Принимать нужно `ItemDto`, а не `Item`
- Валидация нужна не только при создании

Rat-In-Hat replied:
Поменял код стайл, все параметры теперь будут как в вашем примере. 
Хорошо, на вход сделаю также DTO , но вот валидацию на update увы сделать не получится. При обновлении на вход может прийти DTO, у которого нет всех полей, например только описание, валидация его просто не пропустит дальше. Поэтому проверка организована уже дальше внутри кода,  чтобы все таки обновить объект.

FanaticFiz commented on src/main/java/ru/practicum/shareit/item/controller/ItemControllerExceptionHandler.java:
```diff
@@ -0,0 +1,46 @@
+                .build();
+    }
+
+    @ExceptionHandler(value = {UserNotFoundException.class})
```
Мы не создаём классы исключений под каждую сущность...
Создаётся один общий класс исключений.
Бросая его, ты в сообщении будешь писать, что именно произошло.

Ты замучаешься создавать классы под каждую сущность отдельно, в большом проекте.
Класс ExceptionHandler так же должен быть один.

Rat-In-Hat replied:
Согл, сделаю, просто по ТЗ понял что в каждом пакете свое.

FanaticFiz commented on src/main/java/ru/practicum/shareit/user/model/User.java:
```diff
@@ -0,0 +1,23 @@
+public class User {
+    @Positive
+    private Long id;
+    @NotBlank(message = "The name cannot be empty")
```
Валидация должна быть у `dto`. 
Эта сущность это модель для персистентности, тут будут другие аннотации, но не в этой итерации.

Rat-In-Hat replied:
Правильно понял, что у классов User и тп не должно быть валидации? Почему если так?

В любом случае уберу, ведь на вход теперь и так поступает ДТО.

FanaticFiz replied:
Ну частично я попытался сразу пояснить. 
Ну а так просто потому что валидируем мы dto, которая потом мапится в персистентную сущность(entity) `UserDto -> User`. Зачем нам одинаковые проверки в двух местах.

Rat-In-Hat replied:
ага, тупой вопрос задал))

FanaticFiz commented on src/main/java/ru/practicum/shareit/user/service/IUserService.java:
```diff
@@ -0,0 +1,18 @@
+import java.util.List;
+
+public interface IUserService {
+    UserDto createUser(User user);
```
Из названия методов желательно убрать название сущности: `user`.
С чем мы работаем понятно из названия класса.

`createUser -> create`, `getUserById -> getById`

Rat-In-Hat replied:
Исправлено

FanaticFiz commented on src/main/java/ru/practicum/shareit/item/storage/IItemStorage.java:
```diff
@@ -0,0 +1,21 @@
+
+public interface IItemStorage {
+
+    Optional<Item> save(Long userId, Item item);
```
`Optional` уместен только для метода `getItemById`.
При создании или обновлении мы или выполним операцию успешно и можно вернуть саму сущность `Item`, или упадем с ошибкой, но она будет перехвачена и обработана.
Поэтому тут нет неопределенности и не нужно `Optional`

И второе, давай в сервисах использовать глагол `get`, а тут `find`.

Rat-In-Hat replied:
Про опшинал понял, сделаю

А вот почему в сервисе гет а  в хранилище файнд? сделаю, просто для понимания мне)

FanaticFiz replied:
Для порядку! ))
Нет никакой такой магии в этом, просто "я художник - я так вижу".

Сейчас ты подключишь фреймворки и библиотеки и увидишь что там в репозитории `find` будешь воспринимать его как родного!  ))

FanaticFiz commented on src/main/java/ru/practicum/shareit/item/storage/impl/ItemStorageInMemory.java:
```diff
@@ -0,0 +1,153 @@
+
+    @Override
+    public Optional<Item> update(Long userId, Long itemId, Item item) {
+        throwIfItemNotFound(itemId);
```
Ты прописал логику в хранилище и это не правильно!
Все проверки должны были бы быть в сервисе.

Rat-In-Hat replied:
Я сделаю в сервисе, но хотел бы понять почему именно так? Просто мне кажется такие проверки логичнее делать в хранилище и там же кидать исключения. Как оно на практике?
throwIfEmailDuplicate оставил в храниоище, согласен с этим? ПОтому что для его реализации нужен второй метод получения юзера по emai.  Все равно на БД перейдем)

Также приходится в сервис внедрять хранилище, это норм? Или  я что то не так сделал.

Rat-In-Hat replied:
throwIfEmailDuplicate  я вроде таки убрал оттуда, добавил доп метод просто

FanaticFiz replied:
"хотел бы понять почему именно так?" - разделение ответственности.
Хранилище ответственно только за работу с базой(ну или где там у нас хранится все).

"throwIfEmailDuplicate оставил в хранилище, согласен с этим?" - нет, это не верно. 
Это относится к логике работы нашего приложения, а что если мы не хотим запрещать дубли?

Да, все верно, использовать в сервисе хранилища - это то что нужно.
В трёх уровненной архитектуре, контроллеры взаимодействуют с сервисами, а сервисы с хранилищами.

FanaticFiz commented on src/main/java/ru/practicum/shareit/item/storage/impl/ItemStorageInMemory.java:
```diff
@@ -0,0 +1,153 @@
+    @Override
+    public Optional<Item> getItemById(Long itemId) {
+        throwIfItemNotFound(itemId);
+        return Optional.of(items.get(itemId));
```
Вот, например, throwIfItemNotFound(itemId) - должно быть в сервисе.
Тут только `return Optional.of(items.get(itemId));`

Rat-In-Hat replied:
сделал.

FanaticFiz commented on src/main/java/ru/practicum/shareit/user/service/impl/UserService.java:
```diff
@@ -0,0 +1,49 @@
+
+    @Override
+    public UserDto createUser(User user) {
+        userStorage.save(user);
```
Вот сервисы у тебя и пустые и ничего и не делают. Потому что все в хранилище.
Логику реализуем тут... проверяем, например, есть ли такая запись - если есть кидаем исключение.

Rat-In-Hat replied:
перенес)

FanaticFiz commented on src/main/java/ru/practicum/shareit/user/service/impl/UserService.java:
```diff
@@ -0,0 +1,49 @@
+    public UserDto getUserById(Long userId) {
+        return userStorage.getUserById(userId)
+                .map(userMapper::toUserDto)
+                .orElse(null);
```
Не возвращай `null` - это плохая практика!
Тем более что бы позаботился и возвращаешь Optional - теперь осталось им только воспользоваться.

После того как ты сходил в хранилище за объектом ты должен понять вернули его или Optional пуст. Если пусть то кидаешь исключение, если нет - возвращаешь.

Rat-In-Hat replied:
исправил

FanaticFiz commented on src/test/java/ru/practicum/shareit/ShareItTests.java:
```diff
@@ -6,8 +6,7 @@
-	void contextLoads() {
-	}
-
+    @Test
```
`pom` файл не менялся.
Напишу коммент в произвольном месте...

1. Хорошей практикой будет указать кодировку
`<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>`
2. Я советую всем соблюдать некий порядок при объявлении зависимостей.
Сначала "главный" `spring` все его зависимости и всё о нём.
Потом по степени "тяжеловестности" остальные зависимости. Ну и у меня в проекте, например, последними идёт еще одна секция наши собственные библиотеки/модули/контракты.

Rat-In-Hat replied:
когда зашел  в pom файл инфаркт чуть не хватил, много там всего однако, хах) добавил .

FanaticFiz commented on src/test/java/ru/practicum/shareit/ShareItTests.java:
```diff
@@ -6,8 +6,7 @@
-
+    @Test
+    void contextLoads() {
+    }
```
### application.properties

И про файл настроек `application.properties`, так как он не менялся, тоже тут напишу...

Я буду топить за то чтобы ты сразу привыкал к написанию файла свойств для приложения в формате `yaml -> application.yml`
Мне трудно сейчас представить проект в котором до сих пор на него не пересели ))

Соответственно - попрошу тебя этому следовать.

Rat-In-Hat replied:
сделал
дейставительно более красиво

FanaticFiz reviewed (changes requested):
Финальный комментарий на Я.Практикуме

Rat-In-Hat committed:
del

удалены поля id у DTO запрос и буккинг, а также ненужные комментарии

Rat-In-Hat committed:
style

параметры метода теперь с новой строки

Rat-In-Hat committed:
ref

теперь контроллеры принимают DTO

Rat-In-Hat committed:
ref

исключения теперь в общем пакете и с общими именами
обработчик исключение теперь общий

Rat-In-Hat committed:
del

удалены аннотации у клссов User and Item

Rat-In-Hat committed:
ref

имена методов в сервесих и хранилишах больше не содержать имена сущностей

Rat-In-Hat committed:
ref

в серисах теперь get, в хранилищах find

Rat-In-Hat committed:
ref

теперь методы возвращают объекты без Optional

Rat-In-Hat committed:
ref

обработка исключений перенесена в слой сервисов у user and item

Rat-In-Hat committed:
feat

добавлен новый метод в UserStorage, который проверяет есть ли юзер с такой почтой или нет

Rat-In-Hat committed:
ref

метод больше не возвращает null

Rat-In-Hat committed:
feat

добавить <sourceEncoding>UTF-8</sourceEncoding> в pom.xml
поменять application.properties на application.yml

Rat-In-Hat committed:
style

code reformat

Rat-In-Hat committed:
fix

POM исправить

FanaticFiz commented on src/main/java/ru/practicum/shareit/exception/IllegalNameOrDescriptionException.java:
```diff
@@ -0,0 +1,7 @@
+package ru.practicum.shareit.exception;
+
+public class IllegalNameOrDescriptionException extends RuntimeException {
```
Вот тут ты всё еще конкретно про два параметра говоришь.
Получается не сможешь этот класс использовать, когда проблемы будут в других местах?

Называй классы исключений в соответствии с `HTTP` статус кодами которые они будут порождать!
Этот класс стоило бы назвать соответствующе: `BadRequestException`

Rat-In-Hat replied:
поправлю)

Rat-In-Hat committed:
ref

переименовать исключение
